src/codegen/sys/lib_.rs

```diff
@@ -389,20 +440,67 @@ fn generate_classes_structs(w: &mut Write, env: &Env, classes: &[&Class]) -> Res
                 w,
                 "{comment}#[repr(C)]\n{comment}pub struct {name}(c_void);\n",
                 comment = comment,
-                name = klass.c_type
+                name = klass.c_type,
+            ));
+            try!(writeln!(
+                w,
+                "impl ::std::fmt::Debug for {name} {{\n\
+                    \tfn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {{\n\
+                        \t\twrite!(f, \"{name} @ {{:?}}\", self as *const _)\n\
+                    \t}}\n\
+                }}\n",
+                name = klass.c_type,
             ));
         } else {
+            let can_generate_fields_debug = can_generate_fields_debug(&klass.fields);
             try!(writeln!(
                 w,
-                "{comment}#[repr(C)]\n{comment}pub struct {name} {{",
+                "{comment}#[repr(C)]\n{debug}{comment}pub struct {name} {{",
                 comment = comment,
-                name = klass.c_type
+                debug = if can_generate_fields_debug { "#[derive(Debug)]\n" } else { "" },
+                name = klass.c_type,
             ));
 
-            for line in lines {
+            for line in &lines {
                 try!(writeln!(w, "{}{}", comment, line));
             }
             try!(writeln!(w, "{}}}\n", comment));
+
+            if !can_generate_fields_debug && comment.is_empty() {
+                let non_commented_fields = lines.iter()
+                                                .filter(|f| !f.contains("//"))
+                                                .count();
+
+                try!(writeln!(
+                    w,
+                    "impl ::std::fmt::Debug for {name} {{\n\
+                        \tfn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {{\n\
+                            \t\twrite!(f, \"{name} @ {{:?}} {{{{ {fields}}}}}\",\n\
+                            \t\t       self as *const _,\n\
+                            {fields_name})\n\
+                        \t}}\n\
+                    }}\n",
+                    name = klass.c_type,
```

Comment:
Sorry, but this unreadable, and seems repeated at minimum twice.

Reply:
True. I'll write a function and maybe comment or create temporary variables in order to make this less horrible.